### ISSUE TITLE
Github API Usage quota message on Build Release Step

### ISSUE THREAD

**Issue by mpguerra** [openshiftio/openshift.io#3077]
When starting a build and clicking "View Log" I could see the following messages in the Jenkins Console output screen:

```
OpenShift Build rhn-support-pguerra/app-test-1-1 from https://github.com/mpguerra/openshiftio-test.git
13:36:43 Connecting to https://api.github.com with no credentials, anonymous access
13:36:43 GitHub API Usage: Current quota has 30 remaining (8 over budget). Next quota of 60 in 36 min. Sleeping for 15 min.
13:39:43 GitHub API Usage: Still sleeping, now only 12 min remaining.
13:42:43 GitHub API Usage: Still sleeping, now only 9 min 35 sec remaining.
13:45:43 GitHub API Usage: Still sleeping, now only 6 min 35 sec remaining.
13:48:43 GitHub API Usage: Still sleeping, now only 3 min 35 sec remaining.
13:51:43 GitHub API Usage: Still sleeping, now only 35 sec remaining.
13:52:19 GitHub API Usage: Current quota has 30 remaining (2 under budget). Next quota of 60 in 20 min
```

A similar message is also seen on the second Build:

```
OpenShift Build rhn-support-pguerra/app-test-1-2 from https://github.com/mpguerra/openshiftio-test.git
13:22:19 Connecting to https://api.github.com with no credentials, anonymous access
13:22:19 GitHub API Usage: Current quota has 45 remaining (2 over budget). Next quota of 60 in 50 min. Sleeping for 6 min 55 sec.
13:25:19 GitHub API Usage: Still sleeping, now only 3 min 55 sec remaining.
13:28:19 GitHub API Usage: Still sleeping, now only 55 sec remaining.
13:29:14 GitHub API Usage: Current quota has 44 remaining (1 under budget). Next quota of 60 in 43 min
Obtained Jenkinsfile from 342d2a11280ebadbb318a362d6c60a6c14a44680
```

**Comment by ldimaggi** [comment]
Multiple people saw this error during the April 12 'click party' - is this caused by all the jenkins requests/jobs being through a single Jenkins user?

**Comment by mmclanerh** [comment]
This is likely something performing un-authenticated calls to GitHub.  This is based on the error message showing a limit of "60" (vs 5000+).

https://developer.github.com/v3/#rate-limiting 

> For unauthenticated requests, the rate limit allows for up to 60 requests per hour. Unauthenticated requests are associated with the originating IP address, and not the user making requests.

**Comment by pbergene** [comment]
@krishnapaparaju leaving a ping here following todays ic.

**Comment by mmclanerh** [comment]
This may be related? At the very least, it's a very similar error -- which points to a problem within Jenkins:

https://issues.jenkins-ci.org/browse/JENKINS-44776

**Comment by krishnapaparaju** [comment]
@pbergene Ack. looking into this..

**Comment by miteshvp** [comment]
@pbergene @krishnapaparaju as I said earlier, I could see my github credentials being used while building. Snapshot below
<img width="811" alt="image" src="https://user-images.githubusercontent.com/17736013/38972851-a4f2571e-43c0-11e8-82b2-79186c09bf15.png">

**Comment by kbsingh** [comment]
@miteshvp the log above clearly stats anonymous access; so if you can confirm that every github access made through your run infact used the creds, it might not be the same thing

**Comment by miteshvp** [comment]
@kbsingh - I can confirm every build used my git creds correctly. Maybe a re-validation might help here.

**Comment by jfchevrette** [comment]
Where is the jenkins-slave getting the github user token from? From what I can see the slave pods appear to mount a secret called **jenkins-hub-api-token** and at this time the value from that secret is **null** for all tenants.

**Comment by miteshvp** [comment]
It picks it from secret `cd-github` under user namespace

**Comment by jaseemabid** [comment]
## Updates 

1. The error seems to be reported by yet another [plugin](https://github.com/jenkinsci/github-branch-source-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java#L532) we use with practically 0 lines of docs in readme.

**Comment by pbergene** [comment]
It seems to be a blocker for any sort of concurrent use as the anonymous users on the same node count towards a very limited quota. This needs to be sorted as we expect concurrent usage to increase.

**Comment by jaseemabid** [comment]
@pbergene Do you know the flow that causes users to be anonymous?

**Comment by pbergene** [comment]
@jaseemabid see log in first post.

**Comment by krishnapaparaju** [comment]
@animuk @gastaldi need your help to troubleshoot GitHub tokens not being populated / misplaced for booster applications. This problem showed up while concurrent users (around 30) were trying to use Launcher. I am not sure if this is related to one replica currently being used for Launcher backend. @jaseemabid would be great if we can have a check for 'avoiding Github interaction' incase 'anonymous user / access'.  Lets stick with SEV-1.

cc @GeorgeActon

**Comment by gastaldi** [comment]
@krishnapaparaju I don't believe the number of replicas used in the Launcher backend should cause this.

**Comment by gastaldi** [comment]
Launcher calls Jenkins to create the github credentials after the repository is created here:

https://github.com/fabric8-launcher/launcher-backend/blob/1e6975e5e6cee6d8f9b8cf0ce102108d9fae8798/addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioMissionControl.java#L113-L114

Could it be that Jenkins somehow scans the repositories created in an organization before that happens? In this case I think it would be safer to first create the Jenkins credentials and then the repository

**Comment by sthaha** [comment]
@gastaldi even in that case isn't the build is triggered  after the credentials are mounted - https://github.com/fabric8-launcher/launcher-backend/blob/1e6975e5e6cee6d8f9b8cf0ce102108d9fae8798/addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioMissionControl.java#L127  ?

**Comment by gastaldi** [comment]
@staha yes, I think it's something else beyond the scope of Launcher itself

**Comment by sthaha** [comment]
@mpguerra  I tried reproducing this but I am unable to. Are you still able to reproduce this? If so, I want to try inspecting the state of Jenkins pod and also see if the credential is configured properly.

**Comment by sthaha** [comment]
@pbergene @gastaldi I think was @jaseemabid is trying to point to is - https://github.com/jenkinsci/github-branch-source-plugin/blob/a507a73a223ab89fa8051c5b58a64a30953e35b0/src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java#L447 which prints the message - `Connecting to <gh> ...`

**Comment by sthaha** [comment]
### Quickfix

As a quickfix to this issue I am thinking of failing early at https://github.com/fabric8io/fabric8-pipeline-library/blob/0bff0f50a2af9a0dcc4d2c4abb28d1eafcbfdfb9/vars/mavenCI.groovy#L14 if `cd-github` credential is blank/null so that the IP isn't blocked/rate-limited. 
WDYT ?

/cc @pbergene @jaseemabid @gastaldi @sbose78

**Comment by sbose78** [comment]
That's a good place to start, Sunil.

**Comment by mmclanerh** [comment]
For what it's worth, I also didn't produce this in my own account (using experimental features enabled).  Some criteria around when this was seen were:  there was a larger number of concurrency user activity and all users were using the experimental features-enabled quickstarts wizard.  

Thinking about what /might/ cause this..   perhaps we had conditions of a) gh api token not getting set properly or jenkins pod not getting restarted properly after token set and b) many then anonymous github requests (due to high activity) coming from single node (same ip), those two piece could have helped to trigger this condition, right? Following this hypothesis, if we can validate the gh api token(s) getting set properly for any tenants we know affected by this (assuming the pods havent been re-deployed), we may be able to prove/disprove that condition and if present also build a timeline of pod start vs secret setting vs pipeline execution.

Perhaps we can quantify this better through scanning the Jenkins tenant logs across the cluster, looking for "with no credentials, anonymous access", if this has happened at least within the retention for when logs are available.

**Comment by jaseemabid** [comment]
@krishnapaparaju @pradeepto Could you drop the SEV1 label as this is not a blocker anymore? Most of us aren't able to reproduce this.

@sthaha The approach makes sense. That's the only code we understand/own as of now. 

@gastaldi TIL that launcher-backend is a 28000 loc in 337 files large Java project. I wasn't aware of the complicated initialization sequence.

**Comment by kpiwko** [comment]
Is there any way how we could make this issue the top priority? It is blocking performance testing scenario.

**Comment by kpiwko** [comment]
@pmacik do you see any concerns with solution above?

**Comment by maxandersen** [comment]
> Is there any way how we could make this issue the top priority? It is blocking performance testing scenario.

can you give list of namespaces / jenkins urls seeing the issue ?  then i'll go look and see what I can find via logs.

**Comment by maxandersen** [comment]
> @kpiwko Is the fix proposed by @sthaha acceptable for you? If we don't find the secrets mounted, we kill the job early rather than eventually get rate limited?

can we please get a PR ASAP for the proposed change ?

**Comment by maxandersen** [comment]
> @sthaha As far as I understand, the github organization plugin starts the scan and starts the git clone, so I don't completely understand how we could fix this in the pipeline library.

got pointers to this in github org plugin ? I would expect that plugin to also be possible to configure to use proper credentials.

**Comment by rupalibehera** [comment]
we cannot drop the SEV-1 label as we cannot reproduce due not having enough load on the system.

**Comment by rupalibehera** [comment]
> Is there any way how we could make this issue the top priority? It is blocking performance testing scenario.

@kpiwko , could you please help us generating load on the system through your performance tests so that we could reproduce this bug ?

**Comment by pbergene** [comment]
My concern here is that the build teams effort here is not going to address the underlying issue which seems to be missing tokens as opposed to there being calls being done without.  There can be multiple reasons this is missing, but it really does look like we have a bug somewhere in this area. I am doing an iteration over the tokens across east-2 users. While this is not yet complete I am seeing missing tokens in a guesstimate of at least 15% of users.  Will post numbers when it's done. 

I am not entirely sure why this issue exactly (but something in the same area is) would be blocking the perf tests - will do a run over these users too to see if I see the same there.  If we're seeing lacking tokens for accounts that were linked by the same process it would be a good clue, pointing perhaps to some race condition.  

TL;DR impact seems very widespread - currently digging deeper.

**Comment by pbergene** [comment]
east-2 numbers are in.  Keep in mind that this is impact / scope / investigation work and there are multiple reasons for this token missing, so lets not jump to any conclusions - but do chase one.

Reasons may include: 
* User never linked GH, intentionally, cancelled process or simply only ever logged in and didn't do anything else
* User deleted token retrospectively
* User never got directed to a token link page after the getting started experience was removed
* Token linking failed setting the secret
* Probably a few other cases.

Anyway. For 1317 users on east-2, 429 users did not have this secret, for whatever reasons.

Getting more recent numbers from 2a.

Need some more eyes on this, please add in anything relevant here or on triage. @alexeykazakov (4/5 accounts affected).

**Comment by pmacik** [comment]
@rupalibehera  Hi, I can help you with the load.

> @kpiwko , could you please help us generating load on the system through your performance tests so that we could reproduce this bug ?

**Comment by maxandersen** [comment]
> @rupalibehera Hi, I can help you with the load.

@pmacik can we get the list of users/namespaces and times where you seen it being reproduced already ? can help us narrow down the search.

**Comment by pmacik** [comment]
@maxandersen I can see it now for user `osio-ci-ee9` on cluster `start-us-east-2a`:
![image](https://user-images.githubusercontent.com/3263627/39242470-6ca64a5e-488a-11e8-97c2-37024931d684.png)

**Comment by maxandersen** [comment]
FYI: on call on this atm; spotted mismatch between old and new launcher that could cause this. will update ASAP.

**Comment by ldimaggi** [comment]
Max - can you elaborate? What mismatch are you seeing?

**Comment by gastaldi** [comment]
https://github.com/fabric8-launcher/launcher-backend/issues/362 is now fixed. Can someone test in prod-preview before I percolate this change to production?

**Comment by gastaldi** [comment]
After https://ci.centos.org/view/Devtools/job/devtools-launcher-backend-generator-build-master/496/ is finished, of course

**Comment by ldimaggi** [comment]
This issue was surfaced in the click party - and also in automated tests - as the users and tests followed a sequence of resetting the user's environment before attempting to create a new space + app with the launcher. It must have been the case than in many instances Jenkins had not started and anonymous access was used.

**Comment by maxandersen** [comment]
> fabric8-launcher/launcher-backend#362 is now fixed. Can someone test in prod-preview before I percolate this change to production?

took a bit to test since I could not access launcher in prod-preview - but when I could (in incognito mode) secretes now gets created but I can't access jenkins.prod-preview.openshift.io - but not sure that is related ?

**Comment by maxandersen** [comment]
@pmacik @ldimaggi  since this is now in prod, can you run a perf tests where you do *not* prime jenkins ? if that no longer reproduces the issue we can close this.

**Comment by maxandersen** [comment]
as discussed on calls today downgrading this from SEV1 to SEV2 since this is not actually in prod but in experiemental feature; but it is a P0 as we'll need this ready very soon.

**Comment by sthaha** [comment]
Just a heads up, I am adding a fix to the pipeline side as well - https://github.com/fabric8io/fabric8-pipeline-library/pull/401 (WIP)

**Comment by sthaha** [comment]
#### Update

I was under the wrong impression that `withCredentials`(used in pipeline) silently pass when credentials are absent but in fact it fails so, we do not require changes to pipeline library itself. What I found was that the following log is printed even before the control reaches pipeline. 

```
OpenShift Build sunil-thaha/wildfly-test-health-check-13 from https://github.com/sthaha/wildfly-test-health-check.git
05:42:44 Connecting to https://api.github.com with no credentials, anonymous access
05:42:44 GitHub API Usage: Current quota has 55 remaining (2 under budget). Next quota of 60 in 59 min
Obtained Jenkinsfile from 6356d55f80f6cacdd604c70605a42a8c3220b9e9
Loading library github.com/sthaha/fabric8-pipeline-library@fail-early
Attempting to resolve fail-early from remote references...
 > git --version # timeout=10
 > git ls-remote -h -t https://github.com/sthaha/fabric8-pipeline-library.git # timeout=10
Found match: refs/heads/fail-early revision 7cc55c3d4e7768bea02fde15a6c37e097788ef8f
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/sthaha/fabric8-pipeline-library.git # timeout=10
Fetching without tags
Fetching upstream changes from https://github.com/sthaha/fabric8-pipeline-library.git
 > git --version # timeout=10
 > git fetch --no-tags --progress https://github.com/sthaha/fabric8-pipeline-library.git +refs/heads/*:refs/remotes/origin/*
Checking out Revision 7cc55c3d4e7768bea02fde15a6c37e097788ef8f (fail-early)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 7cc55c3d4e7768bea02fde15a6c37e097788ef8f
Commit message: "check gh token is null"
 > git rev-list 7cc55c3d4e7768bea02fde15a6c37e097788ef8f # timeout=10


[Pipeline] withCredentials   <<<<< pipeline starts here 
```

There is a good chance that we may hit this issue if `cd-github` credentials are missing for some reason. And the fix may be in a plugin related to handling github-webhooks

**Comment by krishnapaparaju** [comment]
@gastaldi @animuk please confirm if the fix (secret getting synced up with openshift-sync plugin) for testing..

cc @GeorgeActon

**Comment by gastaldi** [comment]
> @gastaldi <https://github.com/gastaldi> @animuk
> <https://github.com/animuk> please confirm if the fix (secret getting
> synced up with openshift-sync plugin) for testing..
>
You mean the fix in
https://github.com/fabric8-launcher/launcher-backend/commit/a91a2bb71419bf99be559d74a776fad6f41a1c84?
If so, yes, that is in production since last week.

**Comment by sbose78** [comment]
>  was under the wrong impression that withCredentials(used in pipeline) silently pass when credentials are absent but in fact it fails so,

Nice observation @sthaha , even though it fails, the build goes on? and there's no way to terminate the build ( from the code we have in our control ) ?

**Comment by animuk** [comment]
@pmacik Could you try above fix from your test bed by making concurrent users whther the issue still there ?

**Comment by sthaha** [comment]
@sbose78 
> even though it fails, the build goes on?

by fail, I mean the build fails.

**Comment by maxandersen** [comment]
The core cause (launcher not setting up secret but instead just talking to jenkins) have been fixed a few days ago. Verified and thus Closing. If tests shows it happening again lets create new issue.
